plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -126,6 +150,34 @@ findPosition pos root = go Nothing root
             startOfRight <- _start . _codeRange_range <$> V.headM right
             if pos < startOfRight then binarySearchPos left else binarySearchPos right
 
+-- | Traverses through the code range and children to a folding ranges
+findFoldingRanges :: CodeRange -> [FoldingRange]
+findFoldingRanges r@(CodeRange _ children _) =
+  let frRoot :: [FoldingRange] = case createFoldingRange r of
+        Just x  -> [x]
+        Nothing -> []
+
+      frChildren :: [FoldingRange] = concat $ V.toList $ fmap findFoldingRanges children
+   in frRoot ++ frChildren
+
+-- | Parses code range to folding range
+createFoldingRange :: CodeRange -> Maybe FoldingRange
+createFoldingRange (CodeRange (Range (Position lineStart charStart) (Position lineEnd charEnd)) _ ck) = do
+    let frk = crkToFrk ck
+    if lineStart == lineEnd
+        then Nothing
+    else case frk of
+        Just _  -> Just (FoldingRange lineStart (Just charStart) lineEnd (Just charEnd) frk)
+        Nothing -> Nothing
+
+-- Removes all small foldings that start from the same line
+removeDupStartLineFoldings :: [FoldingRange] -> [FoldingRange]
+removeDupStartLineFoldings [] = []
+removeDupStartLineFoldings [x] = [x]
+removeDupStartLineFoldings (frx@(FoldingRange x _ _ _ _):xs@((FoldingRange y _ _ _ _):xs2))
```

Comment:
This isn't a total pattern match, I think? Does GHC not give you a warning here? The pattern is also a bit hard to understand, I'd split it up, maybe just match on `(x:xs)` and then do the rest in let-bindings.

Reply:
This is a total pattern match. 

I'll document how it works in the comments, I guess that should help, mainly understanding the destructuring should help. 

Also, I feel I should rename the x and y to something more self-explanatory. Should I change it?

Reply:
I was also thinking if we can fit this function before we make the folding ranges but that does not seem very convenient. Will that have a significant efficiency advantage? 

It is not convenient because we don't have a place where we can easily judge that these ranges will be duplicates and they start from the same line (like a list of CodeRanges).

I guess the current way is easily maintainable and readable code-wise,

Reply:
I think this way is fine!